orc/orcarm.c

```diff
@@ -1016,3 +998,189 @@ orc_arm64_emit_am (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
+    (((op)&0x3)<<22) | (((imm)&0xfff)<<10) | (((Rn)&0x1f)<<5) | ((Rt)&0x1f))
+
+#define arm64_code_mem_extended(b,op,Rm,opt,S,Rn,Rt) (0xb8200800 | ((((b)==64)&0x1)<<30) | \
+    (((op)&0x3)<<22) | (((Rm)&0x1f)<<16) | (((opt)&0x7)<<13) | (((S)&0x1)<<12) | \
+    (((Rn)&0x1f)<<5) | ((Rt)&0x1f))
+
+void
+orc_arm64_emit_mem (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64Mem opcode,
+    OrcArm64Type type, int opt, int Rt, int Rn, int Rm, orc_uint64 val)
+{
+  orc_uint32 code;
+  orc_uint32 amount;
+  orc_int32 imm;
+
+  int is_signed, extend, S;
+
+  static const char *insn_names[3] = {
+    "str", "ldr"
+  };
+  static const char *extend_names[] = {
+    "ERROR", "ERROR", "uxtw", "lsl",
+    "ERROR", "ERROR", "sxtw", "sxtx"
+  };
+
+  char opt_rn[ARM64_MAX_OP_LEN];
+  char opt_rm[ARM64_MAX_OP_LEN];
+
+  if (opcode >= sizeof(insn_names)/sizeof(insn_names[0])) {
+    ORC_COMPILER_ERROR(p, "unsupported opcode %d", opcode);
+    return;
+  }
+
+  memset (opt_rn, '\x00', ARM64_MAX_OP_LEN);
+  memset (opt_rm, '\x00', ARM64_MAX_OP_LEN);
+
+  switch (type) {
+    case ORC_ARM64_TYPE_IMM:
+      if (opcode != ORC_ARM64_MEM_LDR) {
+        ORC_COMPILER_ERROR(p, "str with immediate is not permitted\n");
+        return;
+      }
+
+      imm = val/4;
+      if (imm == 0) {
+        int label = opt;
+        /** resolve the actual address diff in fixup code */
+        orc_arm_add_fixup (p, label, 2);
+        snprintf (opt_rn, ARM64_MAX_OP_LEN - 1, ", .L%d", label);
+      } else {
+        snprintf (opt_rn, ARM64_MAX_OP_LEN - 1, ", 0x%08x", imm);
+      }
+
+      code = arm64_code_mem_literal (bits, imm, Rt);
+      break;
+    case ORC_ARM64_TYPE_REG:
+      imm = (orc_int32) val;
+      is_signed = opt;
+
+      if (is_signed) { /** signed offset */
+        if (imm > 0xff || imm < -0xff) {
+          ORC_COMPILER_ERROR(p, "simm is out-of-range %x\n", imm);
+          return;
+        }
+
+        if (is_signed == 1) { /** pre index */
+          snprintf (opt_rn, ARM64_MAX_OP_LEN - 1, ", [%s", orc_arm64_reg_name(Rn, bits));
+          snprintf (opt_rm, ARM64_MAX_OP_LEN - 1, ", #%d]!", imm);
+
+          code = arm64_code_mem_signed (bits, opcode, imm, 1, Rn, Rt);
+        } else {              /** post index */
+          snprintf (opt_rn, ARM64_MAX_OP_LEN - 1, ", [%s]", orc_arm64_reg_name(Rn, bits));
+          snprintf (opt_rm, ARM64_MAX_OP_LEN - 1, ", #%d", imm);
+
+          code = arm64_code_mem_signed (bits, opcode, imm, 0, Rn, Rt);
+        }
+      } else {  /** unsigned offset */
+        if (bits == ORC_ARM64_REG_64) {
+          if (imm < 0 || imm > 0xfff * 8)
+            ORC_COMPILER_ERROR(p, "imm is Out-of-range %x\n", imm);
+          imm /= 8;
+        } else {
+          if (imm < 0 || imm > 0xfff * 4)
+            ORC_COMPILER_ERROR(p, "imm is Out-of-range %x\n", imm);
+          imm /= 4;
+        }
+
+        if (imm != 0) {
+          snprintf (opt_rn, ARM64_MAX_OP_LEN - 1, ", [%s", orc_arm64_reg_name(Rn, bits));
```

Comment:
I thought the size of snprintf does not count the terminating null byte; however, the manual says:
```
       The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.
```